src/ZODB/tests/BasicStorage.py

```diff
@@ -385,3 +388,237 @@ def lastInvalidations():
+    # verify storage/Connection for race in between load/open and local invalidations.
+    # https://github.com/zopefoundation/ZEO/issues/166
+    # https://github.com/zopefoundation/ZODB/issues/290
+    @with_high_concurrency
+    def check_race_loadopen_vs_local_invalidate(self):
+        db = DB(self._storage)
+
+		# init initializes the database with two integer objects - obj1/obj2
+        # that are set to 0.
+        def init():
+            transaction.begin()
+            zconn = db.open()
+
+            root = zconn.root()
+            root['obj1'] = MinPO(0)
+            root['obj2'] = MinPO(0)
+
+            transaction.commit()
+            zconn.close()
+
+        # verify accesses obj1/obj2 in a loop and verifies that obj1.value == obj2.value
+        #
+        # access to obj1 is organized to always trigger loading from zstor.
+        # access to obj2 goes through zconn cache and so verifies whether the
+        # cache is not stale.
+        failed = threading.Event()
+        failure = [None]
+        def verify(N):
+            def _verify():
+                transaction.begin()
+                zconn = db.open()
+
+                root = zconn.root()
+                obj1 = root['obj1']
+                obj2 = root['obj2']
+
+                # obj1 - reload it from zstor
+                # obj2 - get it from zconn cache
+                obj1._p_invalidate()
+
+                # both objects must have the same values
+                v1 = obj1.value
+                v2 = obj2.value
+                if v1 != v2:
+                    failure[0] = "verify: obj1.value (%d)  !=  obj2.value (%d)" % (v1, v2)
+                    failed.set()
+
+                transaction.abort() # we did not changed anything; also fails with commit
+                zconn.close()
+
+            try:
+                for i in range(N):
+                    #print('verify.%d' % i)
+                    _verify()
+                    if failed.is_set():
+                        break
+            except:
+                failed.set()
+                raise
+
+
+        # modify changes obj1/obj2 in a loop by doing `objX.value += 1`.
+        #
+        # Since both objects start from 0, the invariant that
+        # `obj1.value == obj2.value` is always preserved.
+        def modify(N):
+            def _modify():
+                transaction.begin()
+                zconn = db.open()
+
+                root = zconn.root()
+                obj1 = root['obj1']
+                obj2 = root['obj2']
+                obj1.value += 1
+                obj2.value += 1
+                assert obj1.value == obj2.value
+
+                transaction.commit()
+                zconn.close()
+
+            try:
+                for i in range(N):
+                    #print('modify.%d' % i)
+                    _modify()
+                    if failed.is_set():
+                        break
+            except:
+                failed.set()
+                raise
```

Comment:
I prefer:
1. move _modify/_verify out of modify/verify
2. merge modify & verify into a single function (you could still have a nested function with code that is common to _modify/_verify)
3. rename _modify/_verify to modify/verify
4. make threading.Thread pass an extra argument that is either modify or verify
5. for the commented print, you can use `__name__` of the passed function

On the other side, no need for threading.Thread to pass N.

Reply:
ok, let's have it this way

> On the other side, no need for threading.Thread to pass N.

except this - I prefer to pass parameters explicitly instead of implicitly referring to shared globals.

￼

Reply:
Since you capitalized it, it's semantically not a variable but a constant.

Reply:
It's just a major parameter.